UI/UX fixes for IR

pavgra opened the pull request:
fixes https://github.com/OHDSI/Atlas/issues/1099

- made IR generations page to be similarly styled as Cohorts one (to improve UI consistency across the app and bring better UX patterns from Cohorts page to IR)
- removed delay between Generation button click and loader being displayed

chrisknoll commented:
Sorry we didn't discuss this first.  I haven't reviewed this PR request so I do not know what you mean by 'made consistent', but if we're going to change any of the generate behavior, here's my suggestion:

First, main problem with current implementation:  when you have many sources, the results screen that shows sources with results is intermingled with sources that do not have results. We have about 20 CDM sources set up in our env, and it's very cumbersome to find a source with results, and generating across multiple sources is also a bit cumbersome.

So, in form-follows-function principle, this is how I would define the generate function requirements:

1. Only sources with results should be displayed.
2. You should be able to batch-execute multiple sources by selecting the sources, then generate.

Item 1 is easy: just show sources that have results. That will eliminate  the problem of figuring out which sources have results, and reduce a lot of clutter.

Item 2 is also fairly simple.  Given the current IR layout:

![image](https://user-images.githubusercontent.com/6818777/52159559-0dbc0b00-2674-11e9-8aad-04a8dcc06a93.png)


You can see the Generate UI element at the top. Currently it is a drop-down that contains the set of sources (not expanded here) but, instead, the generate dropdown can instead be a buton that displays a modal of available sources.  Using a 'click-to-select' checkbox or row select, you can select the set of sources you want to generate on.  After selections are made, you click 'generate' to close the modal and kick off the jobs (or you can cancel if you change your mind). This has the added benefit of not accidentally 'generating' on a source when you didn't intend to (although I understand you can cancel the jobs, it's better not to make the mistake int he first place).

The IR UX experiene was the latest attempt at a streamlined cohort generation process. We never had an opportunity to migrate it to the other sections of the app, and since you have the time to update the generate UX, let's move forward by applying a UX that gives the above advantages instead of moving away from it and making the UX less optimal.

I'm happy to take this on, if you find that you do not have the cycles to do this. but I'm pretty firm against taking a step backwards in UX.

pavgra commented:
@chrisknoll , 

> I haven't reviewed this PR request so I do not know what you mean

Doesn't seem like a constructive approach when it is followed not by some questions but by a waterfall of some assumptions. Anyway, I returned back the existed behavior and going to migrate the same approach to cohorts (if it doesn't occur that the required amount of time is infeasible)

pavgra commented:
![image](https://user-images.githubusercontent.com/10010071/52385210-a248b380-2a4e-11e9-9215-0cee6db14399.png)

![image](https://user-images.githubusercontent.com/10010071/52385235-ae347580-2a4e-11e9-8b32-fd3ce5fe0077.png)

![image](https://user-images.githubusercontent.com/10010071/52385256-ba203780-2a4e-11e9-9a8e-6e82c3ec05ee.png)

![image](https://user-images.githubusercontent.com/10010071/52385282-ce643480-2a4e-11e9-9035-6bb91965af71.png)

chrisknoll commented:
@pavgra ,
 I ran it locally, the changes look good, there is one strange behavior with the UI layout related to resizin the window.

At this size there's no horizontal scrollbar:
![image](https://user-images.githubusercontent.com/6818777/52432465-71ad5c00-2ad8-11e9-83d1-5ea2775c1980.png)

But at this resolution:
![image](https://user-images.githubusercontent.com/6818777/52432483-7ffb7800-2ad8-11e9-8063-23ab70640b5f.png)

the horizontal bar appears.  Oddly, making the window smaller does cause the table to shrink down:
![image](https://user-images.githubusercontent.com/6818777/52432517-9acdec80-2ad8-11e9-96a1-0c32a546dbec.png)

Is this something easily addressed?

chrisknoll commented:
Update:
![image](https://user-images.githubusercontent.com/6818777/52432637-f1d3c180-2ad8-11e9-9952-3520dc1ab699.png)

Seems like the issue stems from the view report button overflowing the table cell but the overall table is not recognizing this and instead of limiting the column size, it just lets the column shrink

pavgra commented:
@chrisknoll , hopefully, columns are better now:

![image](https://user-images.githubusercontent.com/10010071/52437500-0f0e8d00-2ae5-11e9-8a6c-35dbf6a4f3f2.png)

![image](https://user-images.githubusercontent.com/10010071/52437526-1afa4f00-2ae5-11e9-9895-20cfd96a42b2.png)

chrisknoll commented:
Scrollbar is gone, but we get this button wrapping when the window is still fairly wide. I'm gonig to try to adjust the style to apply either a no-wrap or fixed width on the column.  here's what i see:
![image](https://user-images.githubusercontent.com/6818777/52440471-33ba3300-2aec-11e9-997d-c9c212741d10.png)

I'll push up a change in a moment, uless you have a different suggestion.

chrisknoll commented:
Ok this si what I came up with finally:
![image](https://user-images.githubusercontent.com/6818777/52441639-35d1c100-2aef-11e9-945c-8e4c6384aecb.png)


I tried to work with relative sizes and using 'em' for the width specification, but the only way I could lock it in is to use 'px' units (which I think you would disapprove of?)  Anyways, I pushed and you can see for yourself.

pavgra commented:
@chrisknoll , if it works - good. I will push changes if smth better comes into my mind.

What do you think of such Select sources popup?
![image](https://user-images.githubusercontent.com/10010071/52444556-ae884b80-2af6-11e9-8817-beab55b8215a.png)

chrisknoll commented:
Looks very good. Way nicer.  I'm concerned about the select all, but I can worry about that later. What we've seen on our env is that sometimes people like to see cohorts and CC's run across all the souces, which then crushes our databases...so a select all makes it easier to do multi selects but in our env, it's highly unlikely anyone should want to cick taht.

chrisknoll commented:
When using the modal and making selections (in IR), get these errors when clicking on checkboxes (and I do not think this was happening before):

![image](https://user-images.githubusercontent.com/6818777/52457367-5f0e4380-2b27-11e9-8922-e16ce1ae02e0.png)

chrisknoll commented:
I believe the issue is how the `sources` computed is being used in the datatable.

```
this.sources = ko.computed(() => params.sources().map(s => ({ ...s, 
  selected: this.selectedSources().includes(s.key) })));
```
According to the `Renderers.renderCheckbox` function:

```
	function renderCheckbox(field) {
		return '<span data-bind="click: function(d) { d.' + field + '(!d.' + field + '()); } , css: { selected: ' + field + '}" class="fa fa-check"></span>';
	}
```

it is assuming that the 'selected' field in the collection is an observable, that when clicked will toggle the observable's value to true/false. 

I'm making changes now...as i'm looking deeper, there's some confusion about where there's a field 'selected' to maintain the selections, vs a list of 'selected' sources...so I hope to resolve all this shortly, and I will push up some changes.

chrisknoll commented:
Ok, I made some changes, and hope this follows your mental model.  At one point, I was confused a bit as to why the select-sources-btn was being put in each of the results rows, since in the results rows, is the result of a single source, and so no source selection would be necessary in this case (it would just call the execute method for the source).

But, I worked through it, and while I understood the idea of reusing a single 'generate' button for the single-source and multi-source case, i don't think I would have added this complexity into the component: I think where we just know we have a single source and we want to 're-run', I'd just have put a button to re-run that calls execute().  This would remove the need for a 'wasGenerated' parameter (which, the purpose of the component is to push a button, show a list of sources, and then take action for each selected source...it doesn't have to be 'generate').   And then also remove the logic within the 'source selector' where if you have just a single source, it should automatically call the generate method.  I think this would simplify the component into having a single purpose, but I acknowledge I may not see the bigger picture to your overall design.

If you could, can you please pull this branch down and run through your scenerios and make sure I didn't break some edge-case that I didn't anticipate?  If you are good, then we'll merge. I will apply my approval now, and you can decide how to proceed.

chrisknoll commented on js/pages/incidence-rates/components/iranalysis/components/results.html:
```diff
@@ -1,66 +1,105 @@
+					</td>
 				<!-- /ko -->
-				<!-- ko ifnot: info().executionInfo.isValid -->
-				<td colspan="4"><div class="overflow-ellipsis" style="width:345px" data-bind="text: 'Error: ' + info().executionInfo.message, attr:{title: info().executionInfo.message}"></div></td>
```
We removed the columns that store the execution error message, is this no longer being displayed in this revision?

pavgra commented:
@chrisknoll , I've re-pushed fix for console exceptions since my diff is smaller in times. I will integrate the job cancel execution capabilities now (as back-end was merged earlier today). I will also look into the single button usage question - I agree that the complexity is higher and may be I went too far

pavgra commented:
- Added ability to cancel IR:

![image](https://user-images.githubusercontent.com/10010071/52507658-2c645980-2bc0-11e9-9748-7d39148306f0.png)

- Added ability to open popup with the error message if execution failed:

![image](https://user-images.githubusercontent.com/10010071/52507689-4736ce00-2bc0-11e9-93a0-8b7b174f8db4.png)

requires https://github.com/OHDSI/WebAPI/pull/809 to work properly

pavgra commented:
@chrisknoll , thank you for the help. Do we think we are ready for the final merge now?

pavgra commented:
@chrisknoll , would you mind approving the related WebAPI PR?
https://github.com/OHDSI/WebAPI/pull/809
